src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java

```diff
@@ -218,19 +227,32 @@ public synchronized void stop() {
 
     MongoCursor<BsonDocument> createCursor(final MongoSourceConfig cfg, final MongoClient mongoClient) {
         LOGGER.debug("Creating a MongoCursor");
-        ChangeStreamIterable<Document> changeStream = getChangeStreamIterable(cfg, mongoClient);
-        BsonDocument resumeAfter = getResumeAfter(cfg);
-        if (resumeAfter != null) {
-            LOGGER.info("Resuming the change stream at the previous offset");
-            changeStream.resumeAfter(resumeAfter);
-            cachedResumeAfter = null;
+        BsonDocument resumeToken = getResumeToken(cfg);
+        return tryCreateCursor(cfg, mongoClient, resumeToken)
+                .orElseGet(() -> resumeToken == null ? null : tryCreateCursor(cfg, mongoClient, null).orElse(null));
+    }
+
+    private Optional<MongoCursor<BsonDocument>> tryCreateCursor(final MongoSourceConfig cfg, final MongoClient mongoClient,
+                                                                final BsonDocument resumeToken) {
+        try {
+            ChangeStreamIterable<Document> changeStreamIterable = getChangeStreamIterable(cfg, mongoClient);
+            if (resumeToken != null && supportsStartAfter) {
```

Comment:
Only use startAfter if there is a resumeToken and the server supports startAfter.

Reply:
Is the effect the same if the server doesn't support startAfter and only supports resumeAfter?  And if so, do we need both conditions?

Reply:
The scenarios we have to support are dropping the db/collection, dropping the db/collection while only watching certain operationTypes and restarting the connector.

We always try to use `startAfter` if possible - if the last token we saw was an invalidate - the cursor will be valid and can continue to produce events. If not we use `resumeAfter` if the cursor wasn't invalidated, failing that we use a new cursor.

The error scenarios we handle are: error code 260 - invalidated cursor and doesn't support startAfter.  The 260 error occurs when `resumeAfter` is tried to be used and the resumeToken is an invalidate event.

Why prefer startAfter?  It can handle restarting on an invalidate event and therefore capture all data after the drop.  For 4.0 servers a new cursor has to be used and that will only monitor changes after the cursor was created.

Reply:
Ah, that makes sense.